design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,181 @@
+
+1. **Persist UUID in the VM’s Spec Field**  
+   **Description:** The UUID would be generated upon the VM’s first boot and stored in the VM’s spec field.
+
+   **Pros:**
+   - Straightforward to implement.
+   - Avoids adding a new API field, fully backward compatible.
+
+   **Cons:**
+   - This approach "abuses" the spec by storing data that does not represent the user's desired state.
+   - Could lead to cluttered and harder-to-read VM definitions.
+
+
+2. **Generate UUID via Webhook and Store in the Spec Field**  
+   **Description:** If no UUID is specified by the user, a webhook generates a random UUID and assigns it to the VM’s spec field.
+
+   **Pros:**
+   - Similar to the first option (no new API field, backward compatible).
+
+   **Cons:**
+   - Adds complexity by involving webhooks, which could slow down performance if numerous VMs start simultaneously.
+   - May potentially bottleneck `virt-api`.
+
+
+3. **Add a Firmware UUID Field to VM Status**  
+   **Description:** Store the generated UUID within the VM’s status.
+
+   **Pros:**
+   - Keeps the spec clean.
+   - Focuses on VM status for generated information.
+
+   **Cons:**
+   - Could introduce a new API field, affecting long-term maintenance and adding to potential API bloat.
+
+
+4. **Upgrade-Specific Persistence of Firmware.UUID**  
+   **Description:** Just before KubeVirt is upgraded, persist the `Firmware.UUID` of existing VMs in `vm.spec` 
+   After the upgrade, any VM without a `Firmware.UUID` in its spec will be considered new.
+   If `vm.spec.template.spec.Firmware.UUID` is not defined, the system defaults to using `vm.metadata.uid`.
+
+   **Pros:**
+   - The upgrade-specific code can be removed after two or three releases.
+   - Simple logic after the upgrade is completed.
+   - Limited disturbance to git-ops workflows, affecting only pre-existing VMs with the current method.
+
+   **Cons:**
+   - Requires additional code (potentially in `virt-operator`) to handle the upgrade-specific logic.
+
+
+
+### Phase 2: Transition to Universally Unique UUID Generation
+
+In the second phase, the approach for generating the firmware UUID will be modified to ensure global uniqueness.
+This transition could include changing the UUID generation to consider both the VM name and namespace,
+or adopting a pseudo-random hash to guarantee uniqueness across clusters. The solutions for this phase are outlined below:
+
+1. **Use Namespace + Name for UUID Generation**  
+   **Description:** Generate the UUID by combining the VM name and namespace, ensuring uniqueness within the cluster.
+
+   **Pros:**
+   - Simple and straightforward to implement.
+   - Avoids reliance on an external generator for randomness.
+
+   **Cons:**
```

Comment:
This is not acceptable, as it is not universally unique. If you want to consider it, please state so as a "con".

Reply:
fixed